### DISCUSSION TITLE
Type hinting for script events

### DISCUSSION THREAD

**Discussion by uuf6429** [composer/composer#11101] [Q&A / Support] [answered]
Hey!

I'm writing a composer script, however the IDE lacks the event object class types:
<img src="https://user-images.githubusercontent.com/230049/194224814-d6de922c-b350-4e67-8c66-a6c7c9900725.png" width="300" />
Maybe it could make sense to expose these as an interface in a separate composer package (which I can include in my composer.json)?
```json
"require-dev": {
    "composer/script-events": "*"       // hypothetical package
}
```

Or if there such a thing already, maybe it could be mentioned in [the documentation](https://getcomposer.org/doc/articles/scripts.md)?

**Comment by stof** [comment, marked as answer]
> Maybe it could make sense to expose these as an interface in a separate composer package (which I can include in my composer.json)?

this would be bad because you could have version mismatches between your local version and the version in the composer phar.

**Reply by uuf6429** [reply]
True, and I can't see a lot of options.

Perhaps one possibility could be to have some special behaviour inside composer that checks and ensures the interface package it was built with also matches the one in the project json/lock? Maybe similar to the plugin api stuff.

How is this otherwise solved? I'd rather not be coding blindly and requiring composer within the project has the same drawback you mentioned.

**Reply by stof** [reply]
The plugin-api stuff does not involve having 2 versions of the code which might get loaded.

Having a *dev* requirement on composer can have the same drawback, but limited to contributors of your plugin rather than users of it.
